src/clr-angular/utils/i18n/common-strings.service.ts

```diff
@@ -3,65 +3,30 @@
  * This software is released under MIT license.
  * The full license information can be found in LICENSE in the root directory of this project.
  */
-import { SkipSelf, Optional, InjectableProvider, forwardRef } from '@angular/core';
+import { Injectable } from '@angular/core';
 
 import { ClrCommonStrings } from './common-strings.interface';
+import { commonStringsDefault } from './common-strings.default';
 
-// @TODO Put the Required type back in when our minimumly supported version of Angular uses
-// TS 2.8 or greater (should be Angular 7)
-// export class ClrCommonStringsService implements Required<ClrCommonStrings> {
-export class ClrCommonStringsService implements ClrCommonStrings {
-  open = 'Open';
-  close = 'Close';
-  show = 'Show';
-  hide = 'Hide';
-  expand = 'Expand';
-  collapse = 'Collapse';
-  more = 'More';
-  select = 'Select';
-  selectAll = 'Select All';
-  previous = 'Previous';
-  next = 'Next';
-  current = 'Jump to current';
-  info = 'Info';
-  success = 'Success';
-  warning = 'Warning';
-  danger = 'Error';
-  rowActions = 'Available actions';
-  pickColumns = 'Show or hide columns';
-  showColumns = 'Show Columns';
-  sortColumn = 'Sort Column';
-  firstPage = 'First Page';
-  lastPage = 'Last Page';
-  nextPage = 'Next Page';
-  previousPage = 'Previous Page';
-  currentPage = 'Current Page';
-  totalPages = 'Total Pages';
-  minValue = 'Min value';
-  maxValue = 'Max value';
-  modalContentStart = 'Beginning of Modal Content';
-  modalContentEnd = 'End of Modal Content';
-  showColumnsMenuDescription = 'Show or hide columns menu';
-  allColumnsSelected = 'All columns selected';
-  signpostToggle = 'Signpost Toggle';
-  signpostClose = 'Signpost Close';
-  loading = 'Loading';
-  singleSelectionAriaLabel = 'Single selection header';
-  singleActionableAriaLabel = 'Single actionable header';
-  detailExpandableAriaLabel = 'Toggle more row content';
-}
+@Injectable({
+  providedIn: 'root',
+})
+export class ClrCommonStringsService {
+  private strings: ClrCommonStrings = commonStringsDefault;
 
-export function commonStringsFactory(existing?: ClrCommonStrings): ClrCommonStrings {
-  const defaults = new ClrCommonStringsService();
-  if (existing) {
-    return { ...defaults, ...existing };
+  localize(overrides: ClrCommonStrings) {
+    this.strings = { ...this.strings, ...overrides };
   }
-  return defaults;
-}
 
-export const COMMON_STRINGS_PROVIDER: InjectableProvider = {
-  useFactory: commonStringsFactory,
-  // We have a circular dependency for now, we can address it later once these
-  // tree-shakeable providers have proper documentation.
-  deps: [[new Optional(), new SkipSelf(), forwardRef(() => ClrCommonStrings)]],
-};
+  get(key: keyof ClrCommonStrings, tokens: { [key: string]: string } = {}) {
```

Comment:
I think this is too dynamic. If I make a typo in the key its not caught by angular's language service or typescript in the build which leads to a silent error. Something like this will get a better dev experience and type safety in the templates.

```typescript
export class ClrCommonStringsService {
  private _strings = commonStringsDefault;

  get strings(): Readonly<ClrCommonStrings> {
    return this._strings;
  }

  localize(overrides: ClrCommonStrings) {
    this._strings = { ...this.strings, ...overrides };
  }
...
}
```

![Screen Shot 2019-08-01 at 4 17 14 PM](https://user-images.githubusercontent.com/2021067/62328110-d76f1500-b477-11e9-93a8-e0bd5e69aa65.png)

If you update the strings to be public and readonly then they can be read but only set through the localize method. The `get` would still be used but would really just be a find replace method now since the value is passed in.

![Screen Shot 2019-08-01 at 4 21 44 PM](https://user-images.githubusercontent.com/2021067/62328338-7eec4780-b478-11e9-8ee4-25bc3d4e0bac.png)

Reply:
I pushed changes for this